pkg/ipam-controller/controllers/ippool/ippool.go

```diff
@@ -0,0 +1,168 @@
+	"reflect"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	v1helper "k8s.io/component-helpers/scheduling/corev1"
+
+	ipamv1alpha1 "github.com/Mellanox/nvidia-k8s-ipam/api/v1alpha1"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/allocator"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/config"
+)
+
+// IPPoolReconciler reconciles Pool objects
+type IPPoolReconciler struct {
+	PoolsNamespace string
+	NodeEventCh    chan event.GenericEvent
+	client.Client
+	Scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// Reconcile contains logic to sync IPPool objects
+func (r *IPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLog := log.FromContext(ctx)
+	if req.Namespace != r.PoolsNamespace {
+		// this should never happen because of the watcher configuration of the manager from controller-runtime pkg
+		reqLog.Info("Ignoring notification on IPPool from different namespace", "ns", req.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	pool := &ipamv1alpha1.IPPool{}
+	err := r.Client.Get(ctx, req.NamespacedName, pool)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			reqLog.Info("IPPool not found")
+			return ctrl.Result{}, nil
+		}
+		reqLog.Error(err, "failed to get IPPool object from the cache")
+		return ctrl.Result{}, err
+	}
+	reqLog.Info("Notification on IPPool", "name", pool.Name)
+
+	err = config.ValidatePool(pool.Name, pool.Spec.Subnet, pool.Spec.Gateway, pool.Spec.PerNodeBlockSize)
+	if err != nil {
+		return r.handleInvalidSpec(ctx, err, pool)
+	}
+
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList); err != nil {
+		reqLog.Error(err, "failed to list Nodes")
+		return ctrl.Result{}, err
+	}
+	selectedNodes := sets.New[string]()
+	nodeNames := make([]string, 0)
+	for i := range nodeList.Items {
+		node := nodeList.Items[i]
+		if pool.Spec.NodeSelector != nil {
+			match, err := v1helper.MatchNodeSelectorTerms(&node, pool.Spec.NodeSelector)
+			if err != nil {
+				reqLog.Error(err, "failed to match Node", "node", node.Name)
+				continue
+			}
+			if match {
+				selectedNodes.Insert(node.Name)
+				nodeNames = append(nodeNames, node.Name)
+			}
+		} else {
+			selectedNodes.Insert(node.Name)
+			nodeNames = append(nodeNames, node.Name)
+		}
+	}
+
+	sort.Slice(nodeNames, func(i, j int) bool {
+		return nodeNames[i] < nodeNames[j]
+	})
+
+	pa := allocator.CreatePoolAllocatorFromIPPool(ctx, pool, selectedNodes)
```

Comment:
nit: you can drop use of selectedNodes and just pass : `sets.New[string]().Insert(nodeNames...)`

Reply:
Done